From e12603858cae178690c113d73b9d80991c6818da Mon Sep 17 00:00:00 2001
From: Marek Kasik <mkasik@redhat.com>
Date: Wed, 3 Jul 2024 14:49:58 +0200
Subject: [PATCH] frdp-session: Handle freerdp_shall_disconnect() sooner

Handle freerdp_shall_disconnect() at the beginning of update() since it could
be shadowed by another event and the idle_close() wouldn't be called at all.
This fixes issue of not closing the connection when disconnected from inside
of the session.
---
 src/frdp-session.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/src/frdp-session.c b/src/frdp-session.c
index 634f546..1bb5536 100644
--- a/src/frdp-session.c
+++ b/src/frdp-session.c
@@ -714,6 +714,13 @@ update (gpointer user_data)
 
   priv = self->priv;
 
+  if (freerdp_shall_disconnect (priv->freerdp_session)) {
+      priv->update_id = 0;
+      g_idle_add ((GSourceFunc) idle_close, self);
+
+      return FALSE;
+  }
+
   usedHandles = freerdp_get_event_handles (priv->freerdp_session->context,
                                            handles, ARRAYSIZE(handles));
   if (usedHandles == 0) {
@@ -738,13 +745,6 @@ update (gpointer user_data)
     return TRUE;
   }
 
-  if (freerdp_shall_disconnect (priv->freerdp_session)) {
-      g_idle_add ((GSourceFunc) idle_close, self);
-      priv->update_id = 0;
-
-      return FALSE;
-  }
-
   return TRUE;
 }
 
-- 
2.47.1

